### PR TITLE
[RNMobile] Support percentage in columns

### PR DESCRIPTION
**Pull request by lukewalczak**
Fixes / first iteration of: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2197

## Description

That PR introduces the first iteration of the column percentage mechanism. It includes:

* controller to change column width available in column settings and the preview next to the slider:
<details>
Screenshot 2020-07-27 at 14 54 56
</details>

* columns layout picker:

<details>
Screenshot 2020-07-27 at 14 56 14
</details>

## How has this been tested?

1. Open mobile app
2. Add `Columns` block
**3. Expect to see bottom sheet with layout picker**
4. Choose one of the options e.g: `33/33/33`
5. Select the first column
6. Open its settings
7. Change the value e.g. to `90`
**8. Expect that fill level in the preview (to the left of a slider) increased**
9. Close bottom sheet and select the second column
10. Open its settings
11. Press text input (to the right of a slider)
12. Type value with more than 1 decimal e.g: `55.555555` and press `Done`
**13. Expect to have value with 1 decimal**
14. Close bottom sheet and switch to HTML mode
**15. Expect that first column has width equal `90` and the second `55.5`**
16. Save the post and open it on web
**17. Expect to see columns displayed proportionally to set the width values**

## Screenshots 

Kapture 2020-07-27 at 15 12 28

## Types of changes

New feature/functionality - columns percentage

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate. 
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/block-variation-picker/index.native.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import { Icon, close } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import MenuItem from '../inserter/menu-item';
+import styles from './style.scss';
+
+function BlockVariationPicker( { isVisible, onClose, clientId, variations } ) {
+	const createBlocksFromInnerBlocksTemplate = ( innerBlocksTemplate ) => {
```
Comment (dratwas) [inline packages/block-editor/src/components/block-variation-picker/index.native.js:34]: This seems like a function that could be outside of the component :) We don't need to declare a new one on each render.

File: packages/block-editor/src/components/block-variation-picker/index.native.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		onClose();
+	};
+
+	return (
+		<BottomSheet
+			isVisible={ isVisible }
+			onClose={ onClose }
+			title={ __( 'Select a layout' ) }
+			contentStyle={ styles.contentStyle }
+			leftButton={
```
Comment (dratwas) [inline packages/block-editor/src/components/block-variation-picker/index.native.js:69]: Hmmm, I'm wondering if it would be beneficial to use the `useMemo` here. Thanks to that we could avoid creating the leftButton when it's not needed.

```js
const leftButton = useMemo( () => <TouchableWithoutFeedbackonPress={ onClose }> ... </TouchableWithoutFeedbackonPress>, [ onClose, cancelButtonStyle ] );
```

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js [hunk 3]
```diff
@@ -193,6 +210,27 @@ class BottomSheetRangeCell extends Component {
 				}
 			>
 				<View style={ styles.container }>
+					{ columnsPreview && (
```
Comment (dratwas) [inline packages/components/src/mobile/bottom-sheet/range-cell.native.js:213]: I think that it shouldn't be a part of range-cell since it is very specific to the columns block. Could you maybe create a new component where you use range-control and columns preview together instead of adding it here?

File: packages/block-library/src/column/edit.native.js [hunk 4]
```diff
@@ -97,21 +136,36 @@ export default compose( [
...
 
+		const block = getBlock( selectedBlockClientId );
+		const blockOrder = getBlockOrder( parentId );
+
+		const selectedColumnIndex = blockOrder.indexOf( clientId );
+		const columnCount = getBlockCount( parentId );
+		const columnWidths = getBlocks( parentId ).map( ( column ) =>
+			getEffectiveColumnWidth( column, columnCount )
+		);
+		const columnsPreview = { columnWidths, selectedColumnIndex };
```
Comment (dratwas) [inline packages/block-library/src/column/edit.native.js:160]: returning a new reference from withSelect forces a lot of unnecessary renders. Please investigate how we could avoid that here

File: packages/block-library/src/column/edit.native.js [hunk 5]
```diff
@@ -97,21 +136,36 @@ export default compose( [
...
+
 		return {
 			hasChildren,
 			isParentSelected,
 			isSelected,
+			columnCount,
+			columnsPreview,
+			block,
```
Comment (dratwas) [inline packages/block-library/src/column/edit.native.js:168]: I see that you need the `columnCount` and `block` only for determine the `value`, WDYT about something like:
```suggestion
			columnsPreview,
			value: getEffectiveColumnWidth( block, columnCount ),
```

File: packages/components/src/notice/list.native.js [hunk 6]
```diff
@@ -35,7 +35,7 @@ class NoticeList extends Component {
 		}
 
 		return (
-			<View style={ styles.list } key={ notices.lenght }>
+			<View style={ styles.list } key={ notices.length }>
```
Comment (dratwas) [inline packages/components/src/notice/list.native.js:38]: Hmmm? Is it an intentional "change"?

**Comment by lukewalczak** [comment]
**UPDATES**

* Changed `Cancel` button color
* Limit the slider adjustments to whole numbers (while still allowing decimals in the text input)

light | dark
--- | ---
Screenshot 2020-08-03 at 10 29 11 | Screenshot 2020-08-03 at 10 29 06

**Comment by lukewalczak** [comment]
**UPDATES**

* Rename label from `Column percentage` to `Width`

Screenshot 2020-08-04 at 11 27 34

File: packages/block-editor/src/components/block-variation-picker/index.native.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+						<MenuItem
+							item={ v }
+							key={ v.name }
+							onSelect={ () => onVariationSelect( v ) }
+						/>
+					);
+				} ) }
+			</ScrollView>
+			<PanelBody>
+				<UnsupportedFooterControl
```
Comment (dratwas) [inline packages/block-editor/src/components/block-variation-picker/index.native.js:115]: Could we do something with that we use the `UnsupportedFooterControl` for showing the information related to that block? I thought it should be used only for Unsupported features. Maybe it is worth to rename it somehow?

File: packages/block-library/src/column/edit.native.js [hunk 8]
```diff
@@ -57,6 +79,32 @@ function ColumnEdit( {
...
+						rangePreview={
+							<ColumnsPreview
+								columnWidths={ columnWidths }
+								selectedColumnIndex={ selectedColumnIndex }
+							/>
+						}
+					/>
+				</PanelBody>
+				<PanelBody>
+					<UnsupportedFooterControl
```
Comment (dratwas) [inline packages/block-library/src/column/edit.native.js:100]: It seems exactly the same as this one - https://github.com/WordPress/gutenberg/pull/23828/files#diff-ce53a21c0c03b9b6d2e43662fa7eb866R115

Maybe it could be a separate component?

File: packages/block-library/src/columns/edit.native.js [hunk 9]
```diff
@@ -150,6 +156,14 @@ function ColumnsEditContainer( {
 						type="stepper"
 					/>
 				</PanelBody>
+				<PanelBody>
+					<UnsupportedFooterControl
```
Comment (dratwas) [inline packages/block-library/src/columns/edit.native.js:160]: Now i'm quite sure that it should be a separate component :)) 3rd time with the same props

File: packages/components/src/mobile/bottom-sheet/index.native.js [hunk 10]
```diff
@@ -349,6 +342,22 @@ class BottomSheet extends Component {
 
 		const WrapperView = isChildrenScrollable ? View : ScrollView;
 
+		const getHeader = () => (
+			<View>
```
Comment (dratwas) [inline packages/components/src/mobile/bottom-sheet/index.native.js:346]: Could it be a Fragment instead of View?

File: packages/components/src/mobile/bottom-sheet/index.native.js [hunk 11]
```diff
@@ -349,6 +342,22 @@ class BottomSheet extends Component {
...
+				<View style={ styles.bottomSheetHeader }>
+					<View style={ { flex: 1 } }>{ leftButton }</View>
```
Comment (dratwas) [inline packages/components/src/mobile/bottom-sheet/index.native.js:348]: Please move { flex: 1 } to styles :)

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js [hunk 12]
```diff
@@ -59,6 +59,12 @@ class BottomSheetRangeCell extends Component {
 		AppState.removeEventListener( 'change', this.handleChangePixelRatio );
 	}
 
+	toFixed( num ) {
```
Comment (dratwas) [inline packages/components/src/mobile/bottom-sheet/range-cell.native.js:62]: I think that most of the changes in that file could be a part of separate PR where you add `toFixed` prop to the range-cell. This columns PR should contain only the part with `rangePreview`

Comment (lukewalczak) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js:62]: Moved range cell changes into https://github.com/WordPress/gutenberg/pull/24379

File: packages/block-editor/src/components/block-variation-picker/index.native.js (hunk 7 above)
Comment (lukewalczak) [reply packages/block-editor/src/components/block-variation-picker/index.native.js:115]: I agree that name is not descriptive any more and it should be rather named `FooterMessageControl`, however, extracting it to the separate component has less sense for me since:

1. it's a temporary message since we don't display columns proportionally yet,
2. the same code is used in 3 places, but there is no common place for a separated component.

Comment (lukewalczak) [reply packages/block-editor/src/components/block-variation-picker/index.native.js:115]: https://github.com/WordPress/gutenberg/pull/24399

File: packages/block-library/src/column/edit.native.js [hunk 13]
```diff
@@ -57,6 +79,32 @@ function ColumnEdit( {
...
+							/>
+						}
+					/>
+				</PanelBody>
+				<PanelBody>
+					<FooterMessageControl
+						label={ __(
+							'Note: Column layout may vary between themes and screen sizes'
+						) }
+						textAlign="center"
```
Comment (dratwas) [inline packages/block-library/src/column/edit.native.js:104]: WDYT about setting `center` as a default one? We could remove it since we use only 'center' right now

**Review by dratwas — APPROVED** [review]
